src/compressionworker.ts

```diff
@@ -0,0 +1,97 @@
+  onError?: (e: ErrorEvent) => void
+): Worker {
+  let worker = null;
+  if (Worker) {
+    let workerUrl = createWorkerUrl(sampleEnvelope);
+    worker = new Worker(workerUrl);
+    worker.onmessage = onMessage || null;
+    worker.onerror = onError || null;
+  }
+  return worker;
+}
+
+function workerContext() {
+  let workerGlobalScope = self as any;
+  let compress = workerGlobalScope.compress;
+  let config = workerGlobalScope.config;
+  let envelope: IEnvelope = workerGlobalScope.envelope;
+  let nextPayload: string[] = [];
+  let nextPayloadLength = 0;
+  let sequence = 0;
+
+  // Edge case: Don't schedule next upload for XhrError instrumentation events
+  // This helps us avoid the infinite loop in the case when all requests fail (e.g. dropped internet connection)
+  // Infinite loop comes from sending instrumentation about failing to deliver previous delivery failure instrumentation.
+  let nextPayloadIsSingleXhrErrorEvent: boolean =  false;
+
+  onmessage = (evt: MessageEvent) => {
+    let message = JSON.parse(evt.data) as IWorkerMessage;
+    switch (message.type) {
+      case WorkerMessageType.AddEvent:
+        let addEventMsg = message as IAddEventMessage;
+        addEvent(addEventMsg.event, addEventMsg.time);
+        break;
+      case WorkerMessageType.ForceUpload:
+        let forceUploadMsg = message as ITimestampedWorkerMessage;
+        uploadNextPayload(forceUploadMsg.time);
+        break;
+      case WorkerMessageType.Terminate:
+        let terminatedMsg = message as ITimestampedWorkerMessage;
+        uploadNextPayload(terminatedMsg.time);
+        self.close();
+      default:
+        break;
+    }
+  };
+
+  function addEvent(event: IEvent, time: number) {
+    let eventStr = JSON.stringify(event);
+    if (nextPayloadLength > 0 && nextPayloadLength + eventStr.length > config.batchLimit) {
+      uploadNextPayload(time);
+    }
+    nextPayload.push(eventStr);
+    nextPayloadLength += eventStr.length;
+    nextPayloadIsSingleXhrErrorEvent = (nextPayload.length === 1 && event.state && event.state.type === Instrumentation.XhrError);
+    if (nextPayloadLength >= config.batchLimit) {
+      uploadNextPayload(time);
+    }
+  }
+
+  function uploadNextPayload(time: number) {
+    if (nextPayloadLength > 0 && !nextPayloadIsSingleXhrErrorEvent) {
+      envelope.sequenceNumber = sequence++;
+      envelope.time = time;
+      let raw = `{"envelope":${JSON.stringify(envelope)},"events":[${nextPayload.join()}]}`;
+      let compressed = compress(raw);
+      nextPayload = [];
+      nextPayloadLength = 0;
+      upload(compressed, raw);
+    }
+  }
+
+  function upload(compressed: string, uncompressed: string) {
+    let message: IUploadMessage = {
+      type: WorkerMessageType.Upload,
+      compressedData: compressed,
+      rawData: uncompressed
+    };
+    workerGlobalScope.postMessage(JSON.stringify(message));
+  }
+}
+
+function createWorkerUrl(sampleEnvelope: IEnvelope): string {
```

Comment:
Essentially, yes. Worker requires a URL in its constructor, pointing to the code, which it will run. This creates a blob, containing a 'bundle' with ALL required worker routines, which would be executed in a completely isolated worker's scope and returns a URL to this Blob. 

---
In reply to: [136483183](https://github.com/Microsoft/clarity-js/pull/28#discussion_r136483183) [](ancestors = 136483183)

Reply:
Could you add a comment that basically include your response above?

---
In reply to: [137143542](https://github.com/Microsoft/clarity-js/pull/28#discussion_r137143542) [](ancestors = 137143542,136483183)

Reply:
Added it above the 'createWorkerUrl' function

---
In reply to: [137241154](https://github.com/Microsoft/clarity-js/pull/28#discussion_r137241154) [](ancestors = 137241154,137143542,136483183)